### PR TITLE
Add index of unspent transaction outputs

### PR DESCRIPTION
**Pull request by joaopaulofonseca**
Implements an index of unspent transaction outputs (UTXOs). 

This approach optimizes `CWallet::AvailableCoins`, by doing some of its work, each time a wallet transaction changes. For instance, using `CWallet::IsMine` earlier for each transaction output, it manages to keep track only of UTXOs related to wallet addresses. 

That brings improvements on RPCs such as `listunspent` and `fundrawtransaction`. Depending on the size of your wallet and number of UTXOs, performance gains may be up to 98%.

This way, instead of going through all the wallet transactions, the search for UTXOs is made only on this updated list of outputs.

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -1431,8 +1431,11 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
             LogPrintf("%s", errorString);
             return InitError(errorString);
         }
+
```
Comment (pedrobranco) [inline src/init.cpp]: Unuseful new line.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -493,6 +495,49 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+    for (unsigned int i = 0; i < wtx->vout.size(); ++i)
+    {
+        const COutPoint outpoint(hash, i);
+        pair<TxSpends::const_iterator, TxSpends::const_iterator> range;
+        range = mapTxSpends.equal_range(outpoint);
+
+        if(range.first == range.second) {
+            isminetype ismine = IsMine(wtx->vout[i]);
+            if(!filterIsMine || ismine) {
+                mapTxUnspents[outpoint] = make_pair(wtx, ismine);
```
Comment (pedrobranco) [inline src/wallet/wallet.cpp]: Missing std namespace.

File: src/init.cpp [hunk 3]
```diff
@@ -1431,8 +1431,11 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
         if (!pwalletMain)
             return false;
+
+        pwalletMain->RefreshUnspents();
```
Comment (promag) [inline src/init.cpp]: Why is this needed? `CWallet::InitLoadWallet` should somehow do this no?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -1650,6 +1699,59 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
 
     {
         LOCK2(cs_main, cs_wallet);
+#if 1
```
Comment (pedrobranco) [inline src/wallet/wallet.cpp]: I think you should remove this directive and the unused code for better reading of the effective changes and prevent code pollution.

File: src/primitives/transaction.h [hunk 5]
```diff
@@ -11,6 +11,9 @@
 #include "serialize.h"
 #include "uint256.h"
 
+class COutPoint;
+std::size_t hash_value(const COutPoint& o);
```
Comment (promag) [inline src/primitives/transaction.h]: Move to after `COutPoint` definition?

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -1650,6 +1699,59 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
...
+            if(coinControl && coinControl->HasSelected() && !coinControl->fAllowOtherInputs && !coinControl->IsSelected(wtxid, outpoint.n))
+                continue;
+
+            isminetype mine = (uit->second).second;
+            vCoins.push_back(COutput(pcoin, outpoint.n, nDepth,
+                                     ((mine & ISMINE_SPENDABLE) != ISMINE_NO) ||
+                                     (coinControl && coinControl->fAllowWatchOnly && (mine & ISMINE_WATCH_SOLVABLE) != ISMINE_NO)));
+        }
+
+        LogPrintf("Unspent count: %d %d\n", vCoins.size(), mapTxUnspents.size());
```
Comment (pedrobranco) [inline src/wallet/wallet.cpp]: IMO this can be removed, because was used only to debugging the execution in DEV right?

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -493,6 +495,49 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+}
+
+void CWallet::RemoveFromUnspents(const COutPoint &outpoint)
+{
+    UnspentsMap::const_iterator uit = mapTxUnspents.find(outpoint);
+    if(uit != mapTxUnspents.end())
+        mapTxUnspents.erase(uit);
+}
+
+void CWallet::RefreshUnspents()
```
Comment (promag) [inline src/wallet/wallet.cpp]: This function is to clear unspents where isMine is false?

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -493,6 +495,49 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+{
+    for (UnspentsMap::iterator uit = mapTxUnspents.begin(); uit != mapTxUnspents.end();)
+    {
+        const COutPoint &outpoint = uit->first;
+        const CWalletTx* wtx = (uit->second).first;
+
+        isminetype isMine = IsMine(wtx->vout[outpoint.n]);
+        mapTxUnspents[outpoint] = std::make_pair(wtx, isMine);
```
Comment (promag) [inline src/wallet/wallet.cpp]: This should be `uit->second.second = isMine;`.

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -1650,6 +1699,59 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
 
     {
         LOCK2(cs_main, cs_wallet);
+#if 1
+        set<const CWalletTx*> rejectedCoins;
```
Comment (promag) [inline src/wallet/wallet.cpp]: Missing `std::`.

Comment (promag) [inline src/wallet/wallet.cpp]: Mixed feelings about this. Maybe we should have unspents indexed by CWalletTx?

File: src/wallet/wallet.h [hunk 10]
```diff
@@ -593,6 +594,15 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
     MasterKeyMap mapMasterKeys;
     unsigned int nMasterKeyMaxID;
 
+    /**
+     * Used to keep track of unspent outpoints
+     */
+    typedef boost::unordered_map<COutPoint, std::pair<const CWalletTx*, isminetype> > UnspentsMap;
```
Comment (promag) [inline src/wallet/wallet.h]: Should be UnspentMap? Check other types.

Comment (promag) [inline src/wallet/wallet.h]: Should we keep also a `boost::unordered_map<CWalletTx*, COutPoint>` to better filter invalid coins?

File: src/wallet/wallet.cpp (hunk 7 above)
Comment (joaopaulofonseca) [inline src/wallet/wallet.cpp]: @promag it's meant for that, but also to update the isMine value. The thing is, when we load a transaction from the DB, we may not have all the addresses loaded at that point. And we wouldn't know if the address is in our wallet or not. This function goes through all the outputs, after the wallet loads completely.

Comment (promag) [inline src/wallet/wallet.cpp]: But only keeps the ones with isMine a true.

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -493,6 +496,50 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+{
+    for (UnspentsMap::iterator uit = mapTxUnspents.begin(); uit != mapTxUnspents.end();)
+    {
+        const COutPoint &outpoint = uit->first;
+        const CWalletTx* wtx = (uit->second).first;
+
+        isminetype isMine = IsMine(wtx->vout[outpoint.n]);
+        uit->second.second = isMine;
+
+        if(!isMine)
```
Comment (promag) [inline src/wallet/wallet.cpp]: Invert condition and switch instructions.

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -493,6 +496,50 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+
+void CWallet::RefreshUnspents()
+{
+    for (UnspentsMap::iterator uit = mapTxUnspents.begin(); uit != mapTxUnspents.end();)
+    {
+        const COutPoint &outpoint = uit->first;
+        const CWalletTx* wtx = (uit->second).first;
+
+        isminetype isMine = IsMine(wtx->vout[outpoint.n]);
+        uit->second.second = isMine;
```
Comment (promag) [inline src/wallet/wallet.cpp]: If we keep only the isMine then:

```
if (IsMine(wtx->vout[outpoint.n])) {
    uit->second.second = true;
    uit++;
}
else  {
    mapTxUnspents.erase(uit++);
}
```

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -1650,33 +1701,55 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
...
-            const CWalletTx* pcoin = &(*it).second;
+            const COutPoint &outpoint = uit->first;
+            const uint256& wtxid = outpoint.hash;
+            const CWalletTx* pcoin = (uit->second).first;
+
+            if(rejectedCoins.find(pcoin) != rejectedCoins.end())
+                continue;
 
             if (!CheckFinalTx(*pcoin))
+            {
```
Comment (promag) [inline src/wallet/wallet.cpp]: This should be in previous line. See conding style. Also review remaining code.

File: src/wallet/wallet.cpp (hunk 12 above)
Comment (joaopaulofonseca) [inline src/wallet/wallet.cpp]: Don't think that would be very precise. In this case, `IsMine()` doesn't return a bool, but a `isminetype`, and that value is used later. 
But I agree with the rest, makes sense to invert the order of the condition and operations.

File: src/wallet/wallet.h (hunk 10 above)
Comment (jpdffonseca) [inline src/wallet/wallet.h]: Not sure what you meant by:
> Should be UnspentMap? Check other types.

File: src/wallet/wallet.cpp (hunk 7 above)
Comment (joaopaulofonseca) [inline src/wallet/wallet.cpp]: Yes, that's true. The ones with addresses that are not on our wallet are removed from this list.

Comment (pedrobranco) [inline src/wallet/wallet.cpp]: @promag keeps the ones that are different from ISMINE_NO.

``` cxx
enum isminetype
{
    ISMINE_NO = 0,
    ISMINE_WATCH_UNSOLVABLE = 1,
    ISMINE_WATCH_SOLVABLE = 2,
    ISMINE_WATCH_ONLY = ISMINE_WATCH_SOLVABLE | ISMINE_WATCH_UNSOLVABLE,
    ISMINE_SPENDABLE = 4,
    ISMINE_ALL = ISMINE_WATCH_ONLY | ISMINE_SPENDABLE
};
```

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -124,13 +124,21 @@ bool CWallet::AddKeyPubKey(const CKey& secret, const CPubKey &pubkey)
...
+        if(!CWalletDB(strWalletFile).WriteKey(pubkey,
                                                  secret.GetPrivKey(),
-                                                 mapKeyMetadata[pubkey.GetID()]);
+                                                 mapKeyMetadata[pubkey.GetID()]))
+        return false;
     }
+
+    // Refresh current UTXOs to check if some of them were already being tracked
+    // but marked as unspendable. Their info will be updated.
+    RefreshUnspents();
```
Comment (joaopaulofonseca) [inline src/wallet/wallet.cpp]: When importing a private key of a previously imported public address, we need to go through the list of UTXOs to check if they were being tracked and marked as unspendable, and update its info.

File: src/wallet/wallet.cpp (hunk 7 above)
Comment (promag) [inline src/wallet/wallet.cpp]: Ah right.

File: src/primitives/transaction.h [hunk 15]
```diff
@@ -50,6 +50,8 @@ class COutPoint
     std::string ToString() const;
 };
 
+std::size_t hash_value(const COutPoint& o);
```
Comment (promag) [inline src/primitives/transaction.h]: I think this won't be accepted. Let's keep it for the moment.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -493,6 +504,48 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+}
+
+void CWallet::RefreshUnspents()
+{
+    for (TxUnspents::iterator uit = mapTxUnspents.begin(); uit != mapTxUnspents.end();) {
+        const COutPoint &outpoint = uit->first;
+        const CWalletTx* wtx = (uit->second).first;
+
+        isminetype isMine = IsMine(wtx->vout[outpoint.n]);
+        if(isMine) {
```
Comment (promag) [inline src/wallet/wallet.cpp]: Insert newline.

Comment (promag) [inline src/wallet/wallet.cpp]: Missing space after `if`.

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -493,6 +504,48 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+            uit->second.second = isMine;
+            uit++;
+        } else {
```
Comment (promag) [inline src/wallet/wallet.cpp]: `else`in newline.

File: src/primitives/transaction.h (hunk 15 above)
Comment (pedrobranco) [inline src/primitives/transaction.h]: we could do without using `boost::unordered_map`, the execution time is almost the same.

Comment (promag) [inline src/primitives/transaction.h]: If that's the case then it's better to make less changes. We could submit other PR to improve.

File: src/wallet/wallet.cpp [hunk 18]
```diff
@@ -1650,33 +1708,50 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
...
     {
         LOCK2(cs_main, cs_wallet);
-        for (map<uint256, CWalletTx>::const_iterator it = mapWallet.begin(); it != mapWallet.end(); ++it)
-        {
-            const uint256& wtxid = it->first;
-            const CWalletTx* pcoin = &(*it).second;
 
-            if (!CheckFinalTx(*pcoin))
+        std::set<const CWalletTx*> rejectedCoins;
+        for (TxUnspents::const_iterator uit = mapTxUnspents.begin(); uit != mapTxUnspents.end(); ++uit) {
```
Comment (promag) [inline src/wallet/wallet.cpp:1735]: Newline.

File: src/wallet/wallet.cpp [hunk 19]
```diff
@@ -1650,33 +1708,50 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
...
+            const COutPoint &outpoint = uit->first;
+            const uint256& wtxid = outpoint.hash;
+            const CWalletTx* pcoin = (uit->second).first;
+
+            if(rejectedCoins.find(pcoin) != rejectedCoins.end())
```
Comment (promag) [inline src/wallet/wallet.cpp]: It may be logical to have this in first place but evaluate if any of the following tests are cheaper and place them before this.

File: src/wallet/wallet.h [hunk 20]
```diff
@@ -593,6 +594,15 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
     MasterKeyMap mapMasterKeys;
     unsigned int nMasterKeyMaxID;
 
+    /**
+     * Used to keep track of unspent outpoints
+     */
+    typedef boost::unordered_map<COutPoint, std::pair<const CWalletTx*, isminetype> > TxUnspents;
```
Comment (promag) [inline src/wallet/wallet.h]: As said before, use std::set.

**Comment by promag** [comment]
Are there tests covering this?

File: src/wallet/wallet.cpp (hunk 19 above)
Comment (promag) [inline src/wallet/wallet.cpp]: Missing space after `if`. Below are other cases.

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (promag) [inline src/wallet/wallet.cpp]: So you mean this is incomplete?

File: src/wallet/wallet.cpp [hunk 21]
```diff
@@ -474,6 +482,9 @@ bool CWallet::IsSpent(const uint256& hash, unsigned int n) const
 
 void CWallet::AddToSpends(const COutPoint& outpoint, const uint256& wtxid)
 {
+    // since this output is being marked as a spend output, we remove it from the unspent cache
+    RemoveFromUnspents(outpoint);
+
     mapTxSpends.insert(make_pair(outpoint, wtxid));
 
     pair<TxSpends::iterator, TxSpends::iterator> range;
```
Comment (promag) [inline src/wallet/wallet.cpp]: Missing `std::`. Check remaining code.

File: src/wallet/wallet.cpp [hunk 22]
```diff
@@ -1650,33 +1708,50 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
...
-        for (map<uint256, CWalletTx>::const_iterator it = mapWallet.begin(); it != mapWallet.end(); ++it)
-        {
-            const uint256& wtxid = it->first;
-            const CWalletTx* pcoin = &(*it).second;
 
-            if (!CheckFinalTx(*pcoin))
+        std::set<const CWalletTx*> rejectedCoins;
+        for (TxUnspents::const_iterator uit = mapTxUnspents.begin(); uit != mapTxUnspents.end(); ++uit) {
+            const COutPoint &outpoint = uit->first;
+            const uint256& wtxid = outpoint.hash;
```
Comment (promag) [inline src/wallet/wallet.cpp]: Is reference symbol `&` near the type of variable name?

File: src/wallet/wallet.cpp [hunk 23]
```diff
@@ -1650,33 +1708,50 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
...
+            int nDepth = pcoin->GetDepthInMainChain();
+            if (nDepth < 0) {
+                rejectedCoins.insert(pcoin);
+                continue;
             }
+
+            if(!(pcoin->vout[outpoint.n].nValue > 0 || fIncludeZeroValue))
+                continue;
+
+            if(IsLockedCoin(wtxid, outpoint.n))
```
Comment (promag) [inline src/wallet/wallet.cpp]: This test is cheaper than `int nDepth = pcoin->GetDepthInMainChain(); if (nDepth < 0)...`

**Comment by promag** [comment]
How is the cache updated when `CWallet::Add*` or `CWallet::RemoveWatchOnly` are called?

File: src/primitives/transaction.h (hunk 15 above)
Comment (joaopaulofonseca) [inline src/primitives/transaction.h]: Yep. We'll proceed with a `std::map`, since the performance difference to a `boost::unordered_map` are not noticeable.

File: src/wallet/wallet.cpp (hunk 22 above)
Comment (joaopaulofonseca) [inline src/wallet/wallet.cpp]: Fixed. It's consistent now.

File: src/wallet/wallet.cpp (hunk 19 above)
Comment (joaopaulofonseca) [inline src/wallet/wallet.cpp]: Moved cheaper tests up. Difference is not that great, but still cheaper ;)

**Comment by joaopaulofonseca** [comment]
@promag There are no tests covering this yet. I'm working on that.

File: src/wallet/wallet.h (hunk 20 above)
Comment (joaopaulofonseca) [inline src/wallet/wallet.h]: You mean `std::map`?

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (joaopaulofonseca) [inline src/wallet/wallet.cpp]: No, not incomplete. This comment is here to explain why we are calling `RefreshUnspents()` in this place. This covers the case where an UTXO was already cached but marked as unspendable due to missing keys. When adding the private key of a previously imported address, we need to go through the list of UTXOs and update them.

**Comment by joaopaulofonseca** [comment]
@promag 

> How is the cache updated when CWallet::Add\* or CWallet::RemoveWatchOnly are called?

Since `CWallet::RemoveWatchOnly(address)` is called internally and only when a private key is imported, that means that we already had the UTXOs for this address being tracked. Given this, we just need to refresh the current cache to update the "solvability" of these UTXOs. This is already being made.

`CWallet::AddWatchOnly(address)` might get a little tricky. IMO we have two possible ways of action here:
- On one hand, we may just import the address, and the next `rescan` will trigger all the transactions and the following behaviours (add their UTXOs to cache, and so on...)
- On the other hand, after importing an address we can search the wallet transactions for UTXOs (that were not being tracked until that point, because the wallet wasn't watching this address) and update the cache with the new UTXOs, with no need for a `rescan`. Although this approach speeds up UTXO detection, IMO it has at least two issues:
  - If we import a batch of addresses, and the cache is updated after each one is imported, we might have performance issues on the whole importation process. 
  - We end up with some new UTXOs (because of the Tx that we already had on our wallet) but without a proper `rescan`, some other unspents will be left out, since we were never aware of them.

WDYT?

**Comment by promag** [comment]
The description should mention the IsMine attribute.

**Comment by promag** [comment]
Please investigate other calls to IsMine in wallet.cpp because they could use the index instead. This will improve rescans for instance.

**Comment by joaopaulofonseca** [comment]
@promag you're right. There are some calls that may be optimized with this changes. Do you think it should be included in this PR? We could submit this first and then work on those improvements after it gets merged. WDYT?

BTW, description was updated. I think the changes are clearer now.

**Comment by promag** [comment]
I said investigate :stuck_out_tongue:. But since this is PR is **Add index of unspent transaction outputs** then I think it could have at least 2 commits: one adding the index and keeping it updated; and other(s) making use of the index.

**Comment by promag** [comment]
Also, let's try to submit this upstream until next wednesday.

**Comment by joaopaulofonseca** [comment]
Ok, that makes sense. I'll split this into 2 commits and then we can push this upstream if you think it's ready. That way we can get start getting some feedback..

**Comment by pedrobranco** [comment]
Please rebase (if already done ignore).

**Comment by joaopaulofonseca** [comment]
@pedrobranco done :+1:

**Comment by jpdffonseca** [comment]
Deprecated by bitcoin PR https://github.com/bitcoin/bitcoin/pull/7823.
